### PR TITLE
Apply dynamodb autoscale

### LINKED ISSUE #536: [Feature] Support Application Autoscale for DynamoDB table read capacity
## Background
[DynamoDB Auto Scaling](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ProvisionedThroughput.html) is a useful feature when DynamoDB IOC table (built by `Threat Intel Downloader`) is processing large number of burst records by `batch_get` from Rule Processor(s). The auto scaling feature will automatically adjust Read capacity based on the traffic. 

Note: AWS DynamoDB Auto Scaling is also support Write Capacity. We don't support Write Capacity auto scaling right now, because,
* The Consumed Write Capacity is related low compared to Read Capacity.
* Write Capacity is more expensive than Read Capacity. With auto scaling it may generate unexpected cost. Right now, it is reasonable to throttle the Write Capacity than auto scaling.

### PR DESCRIPTION
**Pull request by chunyong-lin**
to: @jacknagz @ryandeivert 
cc: @airbnb/streamalert-maintainers
size: medium
resolves #536

## Background
[DynamoDB Auto Scaling](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ProvisionedThroughput.html) is a useful feature when DynamoDB IOC table (built by `Threat Intel Downloader`) is processing large number of burst records by `batch_get` from Rule Processor(s). The auto scaling feature will automatically adjust Read capacity based on the traffic. 

Note: AWS DynamoDB Auto Scaling is also support Write Capacity. We don't support Write Capacity auto scaling right now, because,
* The Consumed Write Capacity is related low compared to Read Capacity.
* Write Capacity is more expensive than Read Capacity. With auto scaling it may generate unexpected cost. Right now, it is reasonable to throttle the Write Capacity than auto scaling. 

## Changes

* Add a new Assume Role for application autoscale service, as well as autoscale policy.
* Autoscale settings are variables and can be set via CLI.
* Correct TTL attribute name.

## Local Testing
* Rule test
```
python manage.py lambda test --processor all
...
StreamAlertCLI [INFO]: (60/60) Successful Tests
StreamAlertCLI [INFO]: (93/93) Alert Tests Passed
StreamAlertCLI [INFO]: Completed
```
* Unit test
```
./tests/scripts/unit_tests.sh
...
-------------------------------------------------------------------------------------
TOTAL                                                    2955     41    99%
----------------------------------------------------------------------
Ran 503 tests in 8.498s

OK
```
## Testing in AWS testing account
* Toggle `autoscale` setting from CLI and `conf/lambda.json`
* Test event 1: 100 times(while loop) batch_get_item with 98 items, where includes 8 hits
  * Table autoscale settings
    * Max read capacity is set to 100
    * Min read capacity is set to 10
    * Utilization is set to 70%
  * After sending batch_get_item, table statistics is
  ```
  aws dynamodb describe-table --table-name my_testing_table_name --query "Table.
  [TableName,TableStatus,ProvisionedThroughput]"
  [
      "my_testing_table_name",
      "ACTIVE",
      {
          "NumberOfDecreasesToday": 0,
          "WriteCapacityUnits": 10,
          "LastIncreaseDateTime": 1513280911.803, (<=2017-12-14 19:48:31)
          "ReadCapacityUnits": 69
      }
  ]
  ```
  * Read Capacity Metrics
test_event1
* Test event 2: 1000 times batch_get_item with 98 items, where includes 8 hits
  * Table autoscale settings
    * Max read capacity is set to **1000**
    * Min read capacity is set to 10
    * Utilization is set to 70%
  * After sending batch_get_item, table statistics is
  ```
  aws dynamodb describe-table --table-name my_testing_table_name --query 
  [
      "my_testing_table_name",
      "ACTIVE",
      {
          "NumberOfDecreasesToday": 0,
          "WriteCapacityUnits": 10,
          "LastIncreaseDateTime": 1513284222.211, (<=2017-12-14 20:43:42)
          "ReadCapacityUnits": 173
      }
  ]
  ```
  * Read Capacity Metrics
test_event3

* Test event 3: Continually sending batch_get_item with 98 items, where includes 8 hits. Launch three processes to mimic three lambda functions doing query to IOC dynamodb table with 50 mins stress.
  * Table autoscale settings
    * Max read capacity is set to **1000**
    * Min read capacity is set to 10
    * Utilization is set to 70%
  * table statistics
  ```
  [
      "chunyong-my_testing_table_name",
      "ACTIVE",
      {
          "NumberOfDecreasesToday": 1,
          "WriteCapacityUnits": 10,
          "LastIncreaseDateTime": 1513290871.606,  (<=2017-12-14 22:34:31)
          "ReadCapacityUnits": 527,
          "LastDecreaseDateTime": 1513285537.587 
      }
  ]
  ```
  * Read Capacity Metrics
test_event4

### CODE REVIEW

File: stream_alert_cli/terraform/threat_intel_downloader.py [hunk 1]
```diff
@@ -53,6 +53,7 @@ def generate_threat_intel_downloader(config):
         'table_wcu': ti_downloader_config.get('table_wcu', '10'),
         'ioc_keys': ti_downloader_config.get('ioc_keys'),
         'ioc_filters': ti_downloader_config.get('ioc_filters'),
-        'ioc_types': ti_downloader_config.get('ioc_types')
+        'ioc_types': ti_downloader_config.get('ioc_types'),
```
Comment (jacknagz) [inline stream_alert_cli/terraform/threat_intel_downloader.py]: do these need default values?

File: terraform/modules/tf_threat_intel_downloader/dynamodb.tf [hunk 2]
```diff
@@ -18,3 +18,80 @@ resource "aws_dynamodb_table" "threat_intel_ioc" {
     Name = "StreamAlert"
   }
 }
+
+// IAM Role: Application autoscalling role
+resource "aws_iam_role" "appautoscaling" {
```
Comment (jacknagz) [inline terraform/modules/tf_threat_intel_downloader/dynamodb.tf]: `appautoscaling` is not super descriptive within the context of StreamAlert

File: terraform/modules/tf_threat_intel_downloader/dynamodb.tf [hunk 3]
```diff
@@ -18,3 +18,80 @@ resource "aws_dynamodb_table" "threat_intel_ioc" {
...
+// IAM Role Policy: Allow appautoscaling IAM role to autoscaling DynamoDB table
+resource "aws_iam_role_policy" "appautoscaling" {
+  name   = "DynamoDBAppAutoscalePolicy"
+  role   = "${aws_iam_role.appautoscaling.id}"
+  policy = "${data.aws_iam_policy_document.appautoscaling.json}"
+}
+
+// IAM Policy Doc: Allow autoscaling IAM role to send alarm to CloudWatch
+// and change table settings for autoscaling.
+data "aws_iam_policy_document" "appautoscaling" {
```
Comment (jacknagz) [inline terraform/modules/tf_threat_intel_downloader/dynamodb.tf]: `appautoscaling` should be swapped for something descriptive to what this policy allows the role to do, like `cloudwatch_write` or something along those lines

File: terraform/modules/tf_threat_intel_downloader/dynamodb.tf [hunk 4]
```diff
@@ -18,3 +18,80 @@ resource "aws_dynamodb_table" "threat_intel_ioc" {
...
+  statement {
```
Comment (jacknagz) [inline terraform/modules/tf_threat_intel_downloader/dynamodb.tf]: respective `dynamodb` and `cloudwatch` permissions should be split into two statements

File: terraform/modules/tf_threat_intel_downloader/dynamodb.tf [hunk 5]
```diff
@@ -18,3 +18,80 @@ resource "aws_dynamodb_table" "threat_intel_ioc" {
...
+
+    resources = [
+      "${aws_dynamodb_table.threat_intel_ioc.arn}",
+      "${aws_cloudwatch_log_group.threat_intel_downloader.arn}",
+    ]
+  }
+}
+
+resource "aws_appautoscaling_target" "dynamodb_table_read_target" {
+  count              = "${var.autoscale ? 1 : 0}"
```
Comment (jacknagz) [inline terraform/modules/tf_threat_intel_downloader/dynamodb.tf]: this `count` should be on all related autoscaling resources (including IAM roles/policies)

File: terraform/modules/tf_threat_intel_downloader/dynamodb.tf [hunk 6]
```diff
@@ -18,3 +18,80 @@ resource "aws_dynamodb_table" "threat_intel_ioc" {
...
+  max_capacity       = 100
+  min_capacity       = 5
+  resource_id        = "${var.prefix}_streamalert_threat_intel_downloader"
```
Comment (jacknagz) [inline terraform/modules/tf_threat_intel_downloader/dynamodb.tf]: I'd recommend interpolating this from the [dyanmodb_table](https://www.terraform.io/docs/providers/aws/r/dynamodb_table.html) resource instead.  This way, if the name ever changes you don't need to update it in two places

File: stream_alert_cli/terraform/threat_intel_downloader.py (hunk 1 above)
Comment (chunyong-lin) [reply stream_alert_cli/terraform/threat_intel_downloader.py]: Yes, the default values are set in the [argparser](https://github.com/airbnb/streamalert/blob/master/manage.py#L1372-L1376)

Comment (jacknagz) [reply stream_alert_cli/terraform/threat_intel_downloader.py]: But that doesn't mean somebody can't modify the config by hand

Comment (chunyong-lin) [reply stream_alert_cli/terraform/threat_intel_downloader.py]: Aha, good point! I will add default values.

File: terraform/modules/tf_threat_intel_downloader/dynamodb.tf (hunk 2 above)
Comment (chunyong-lin) [reply terraform/modules/tf_threat_intel_downloader/dynamodb.tf]: I rename it to `stream_alert_dynamodb_appautoscaling`

File: terraform/modules/tf_threat_intel_downloader/dynamodb.tf (hunk 3 above)
Comment (chunyong-lin) [reply terraform/modules/tf_threat_intel_downloader/dynamodb.tf]: Yes, I renamed the policies to be more descriptive in the newer commit.

File: terraform/modules/tf_threat_intel_downloader/dynamodb.tf (hunk 5 above)
Comment (chunyong-lin) [reply terraform/modules/tf_threat_intel_downloader/dynamodb.tf]: Yes, `count` is applied to all autoscaling resources.
* aws_iam_role
* aws_iam_policy_document
* aws_iam_role_policy
* aws_appautoscaling_target
* aws_appautoscaling_policy

**Comment by chunyong-lin** [comment]
@jacknagz PTAL, I addressed your comments.

File: stream_alert_cli/terraform/threat_intel_downloader.py [hunk 7]
```diff
@@ -51,8 +51,12 @@ def generate_threat_intel_downloader(config):
         'monitoring_sns_topic': dlq_topic,
         'table_rcu': ti_downloader_config.get('table_rcu', '10'),
         'table_wcu': ti_downloader_config.get('table_wcu', '10'),
-        'ioc_keys': ti_downloader_config.get('ioc_keys'),
-        'ioc_filters': ti_downloader_config.get('ioc_filters'),
-        'ioc_types': ti_downloader_config.get('ioc_types')
+        'ioc_keys': ti_downloader_config.get('ioc_keys', ['expiration_ts', 'itype', 'source', 'type', 'value']),
+        'ioc_filters': ti_downloader_config.get('ioc_filters', ['crowdstrike', '@airbnb.com']),
+        'ioc_types': ti_downloader_config.get('ioc_types', ['domain', 'ip', 'md5']),
+        'autoscale': ti_downloader_config.get('autoscale', 'disable'),
```
Comment (jacknagz) [inline stream_alert_cli/terraform/threat_intel_downloader.py]: shouldn't this be a boolean?

**Pull request merged**